src/rgw/store/sfs/sfs_gc.cc

```diff
@@ -0,0 +1,193 @@
+    if (max_objects <= 0) {
+      break;
+    }
+    delete_bucket(bucket_id);
+  }
+}
+
+void SFSGC::delete_objects(const std::string & bucket_id) {
+  sqlite::SQLiteObjects db_objs(store->db_conn);
+  auto objects = db_objs.get_objects(bucket_id);
+  for (auto const& object : objects) {
+    if (max_objects <= 0) {
+      break;
+    }
+    auto obj_instance = std::make_shared<Object>(object.name,
+                                                 object.uuid,
+                                                 true);
+    delete_object(obj_instance);
+  }
+}
+
+void SFSGC::delete_versioned_objects(const std::shared_ptr<Object> & object) {
+  sqlite::SQLiteVersionedObjects db_ver_objs(store->db_conn);
+  auto versions = db_ver_objs.get_versioned_objects(object->path.get_uuid());
+  for (auto const& version : versions) {
+    if (max_objects <= 0) {
+      break;
+    }
+    delete_versioned_object(object, version.id);
+  }
+}
+
+void SFSGC::delete_bucket(const std::string & bucket_id) {
+  // delete the objects of the bucket first
+  delete_objects(bucket_id);
+  if (max_objects > 0) {
+    sqlite::SQLiteBuckets db_buckets(store->db_conn);
+    db_buckets.remove_bucket(bucket_id);
+    lsfs_dout(this, 10) << "Deleting bucket: "
+                        << bucket_id
+                        << dendl;
+    --max_objects;
+  }
+}
+
+void SFSGC::delete_object(const std::shared_ptr<Object> & object) {
+  // delete its versions first
+  delete_versioned_objects(object);
+  if (max_objects > 0) {
+    object->delete_object(store);
+    lsfs_dout(this, 10) << "Deleting object: "
+                        << object->path.get_uuid()
+                        << dendl;
+    --max_objects;
+  }
+}
+
+void SFSGC::delete_versioned_object(const std::shared_ptr<Object> & object,
+                                    uint id) {
+  object->version_id = id;
+  object->delete_object_version(store);
+  lsfs_dout(this, 10) << "Deleting version: ("
+                      << object->path.get_uuid()
+                      << ","
+                      << id
+                      << ")"
+                      << dendl;
+  --max_objects;
+}
+
+SFSGC::GCWorker::GCWorker(const DoutPrefixProvider *_dpp,
+                          CephContext *_cct,
+                          SFSGC *_gc)
+  : dpp(_dpp), cct(_cct), gc(_gc) {
+}
+
+void *SFSGC::GCWorker::entry() {
+  do {
+    utime_t start = ceph_clock_now();
+    lsfs_dout(dpp, 2) << "start" << dendl;
+    int r = gc->process();
+    if (r < 0) {
+      lsfs_dout(dpp, 0)
+         << "ERROR: garbage collection process() returned error r="
+         << r
+         << dendl;
+    }
+    lsfs_dout(dpp, 2) << "stop" << dendl;
+
+    if (gc->going_down())
+      break;
+
+    utime_t end = ceph_clock_now();
+    end -= start;
+    int secs = cct->_conf->rgw_gc_processor_period;
+    secs -= end.sec();
+
+    std::unique_lock locker{lock};
+    cond.wait_for(locker, std::chrono::seconds(secs));
```

Comment:
is this `wait_for()` for filling the remaining amount of time not consumed for the cycle, correct?
what if `secs` is negative because, for any reason: `secs - elapsed` is a negative number?

Reply:
It doesn't wait (tested it)

While I understand the concerns around this... We're talking about seconds precision here.

IMO, the GC should not trigger too often (in ceph defaults to 1h).

And even if we make it trigger every second.... (which I think is not a realistic scenario) if it takes more that 1 second to cleanup... we still have the `rgw_gc_max_objs` parameter to limit the GC... (and possibly another future policy based on elapsed time) and if it's still taking more than 1 second I would be worried about the whole GC, tbh. 

Let's say we configure it to run every 5 mins...
If the GC takes 5 mins to do its job... I would better make the GC delete itself for being completely incompetent. :) or maybe change the `rgw_gc_max_objs` for being crazy high... (or maybe we have thousands of users uploading thousands of objects and buckets and starving the GC thread)

Having said that. I added some code in case the GC is so slow.... to make it wait for 1 period. :)